File: local-bridge/BoardCAD3DMouseBridge.swift
#!/usr/bin/env swift

import Foundation
import IOKit.hid
import Network

private let bridgePort: UInt16 = 8766
private let threeDconnexionVendorID = 0x256F

final class BridgeState {
  private let queue = DispatchQueue(label: "boardcad.3dmouse.state")
  private var axes: [String: Double] = [
    "tx": 0, "ty": 0, "tz": 0,
    "rx": 0, "ry": 0, "rz": 0
  ]
  private var buttons = Set<Int>()
  private var connected = false
  private var deviceName = ""
  private var packetCount = 0
  private var timestamp = Date().timeIntervalSince1970 * 1000

  func setDevice(name: String, connected: Bool) {
    queue.sync {
      self.deviceName = name
      self.connected = connected
      self.timestamp = Date().timeIntervalSince1970 * 1000
      if !connected {
        self.axes.keys.forEach { self.axes[$0] = 0 }
        self.buttons.removeAll()
      }
    }
  }

  func updateAxis(_ key: String, value: Double) {
    queue.sync {
      self.axes[key] = max(-1, min(1, value))
      self.packetCount += 1
      self.timestamp = Date().timeIntervalSince1970 * 1000
    }
  }

  func updateButton(_ index: Int, pressed: Bool) {
    queue.sync {
      if pressed { self.buttons.insert(index) } else { self.buttons.remove(index) }
      self.packetCount += 1
      self.timestamp = Date().timeIntervalSince1970 * 1000
    }
  }

  func stateJSON() -> String {
    queue.sync {
      let payload: [String: Any] = [
        "connected": connected,
        "deviceName": deviceName,
        "timestamp": timestamp,
        "packetCount": packetCount,
        "tx": axes["tx"] ?? 0,
        "ty": axes["ty"] ?? 0,
        "tz": axes["tz"] ?? 0,
        "rx": axes["rx"] ?? 0,
        "ry": axes["ry"] ?? 0,
        "rz": axes["rz"] ?? 0,
        "buttons": Array(buttons).sorted()
      ]
      return encodeJSON(payload)
    }
  }

  func statusJSON() -> String {
    queue.sync {
      let payload: [String: Any] = [
        "ok": true,
        "connected": connected,
        "deviceName": deviceName,
        "timestamp": timestamp,
        "packetCount": packetCount,
        "port": bridgePort
      ]
      return encodeJSON(payload)
    }
  }
}

func encodeJSON(_ payload: [String: Any]) -> String {
  guard let data = try? JSONSerialization.data(withJSONObject: payload, options: [.sortedKeys]),
        let text = String(data: data, encoding: .utf8) else {
    return #"{"ok":false}"#
  }
  return text
}

func normalizeAxis(value: Int, min: Int, max: Int) -> Double {
  if min < 0 && max > 0 {
    let denom = Double(max(abs(min), abs(max)))
    if denom <= 1e-9 { return 0 }
    return Double(value) / denom
  }
  let center = Double(min + max) / 2.0
  let half = Double(max - min) / 2.0
  if half <= 1e-9 { return 0 }
  return (Double(value) - center) / half
}

func axisKey(usagePage: UInt32, usage: UInt32) -> String? {
  guard usagePage == UInt32(kHIDPage_GenericDesktop) else { return nil }
  switch usage {
  case UInt32(kHIDUsage_GD_X): return "tx"
  case UInt32(kHIDUsage_GD_Y): return "ty"
  case UInt32(kHIDUsage_GD_Z): return "tz"
  case UInt32(kHIDUsage_GD_Rx): return "rx"
  case UInt32(kHIDUsage_GD_Ry): return "ry"
  case UInt32(kHIDUsage_GD_Rz): return "rz"
  default: return nil
  }
}

let bridgeState = BridgeState()

final class HIDBridge {
  private let manager: IOHIDManager

  init?() {
    manager = IOHIDManagerCreate(kCFAllocatorDefault, IOOptionBits(kIOHIDOptionsTypeNone))
    let match: [[String: Any]] = [
      [kIOHIDVendorIDKey as String: threeDconnexionVendorID]
    ]
    IOHIDManagerSetDeviceMatchingMultiple(manager, match as CFArray)
    let unmanaged = Unmanaged.passUnretained(self)
    IOHIDManagerRegisterDeviceMatchingCallback(manager, { context, _, _, device in
      guard let context else { return }
      let owner = Unmanaged<HIDBridge>.fromOpaque(context).takeUnretainedValue()
      owner.deviceMatched(device)
    }, unmanaged.toOpaque())
    IOHIDManagerRegisterDeviceRemovalCallback(manager, { _, _, _, device in
      let name = (IOHIDDeviceGetProperty(device, kIOHIDProductKey as CFString) as? String) ?? "3Dconnexion"
      bridgeState.setDevice(name: name, connected: false)
    }, unmanaged.toOpaque())
    IOHIDManagerRegisterInputValueCallback(manager, { _, _, _, value in
      let element = IOHIDValueGetElement(value)
      let usagePage = IOHIDElementGetUsagePage(element)
      let usage = IOHIDElementGetUsage(element)
      if usagePage == UInt32(kHIDPage_Button) {
        let buttonIndex = Int(usage)
        bridgeState.updateButton(buttonIndex, pressed: IOHIDValueGetIntegerValue(value) != 0)
        return
      }
      guard let key = axisKey(usagePage: usagePage, usage: usage) else { return }
      let raw = IOHIDValueGetIntegerValue(value)
      let min = IOHIDElementGetLogicalMin(element)
      let max = IOHIDElementGetLogicalMax(element)
      bridgeState.updateAxis(key, value: normalizeAxis(value: raw, min: min, max: max))
    }, unmanaged.toOpaque())
    IOHIDManagerScheduleWithRunLoop(manager, CFRunLoopGetMain(), CFRunLoopMode.defaultMode.rawValue)
    let result = IOHIDManagerOpen(manager, IOOptionBits(kIOHIDOptionsTypeNone))
    guard result == kIOReturnSuccess else { return nil }
  }

  private func deviceMatched(_ device: IOHIDDevice) {
    let name = (IOHIDDeviceGetProperty(device, kIOHIDProductKey as CFString) as? String) ?? "3Dconnexion"
    bridgeState.setDevice(name: name, connected: true)
  }
}

final class HTTPBridgeServer {
  private let listener: NWListener
  private let queue = DispatchQueue(label: "boardcad.3dmouse.http")

  init?(port: UInt16) {
    guard let port = NWEndpoint.Port(rawValue: port),
          let listener = try? NWListener(using: .tcp, on: port) else { return nil }
    self.listener = listener
    self.listener.newConnectionHandler = { connection in
      connection.start(queue: self.queue)
      self.receive(on: connection)
    }
  }

  func start() {
    listener.start(queue: queue)
    print("BoardCAD 3D mouse bridge listening on http://127.0.0.1:\(bridgePort)")
  }

  private func receive(on connection: NWConnection) {
    connection.receive(minimumIncompleteLength: 1, maximumLength: 8192) { data, _, _, _ in
      let request = String(data: data ?? Data(), encoding: .utf8) ?? ""
      let firstLine = request.split(separator: "\r\n").first.map(String.init) ?? ""
      let parts = firstLine.split(separator: " ")
      let method = parts.count > 0 ? String(parts[0]) : "GET"
      let path = parts.count > 1 ? String(parts[1]) : "/"
      let body: String
      switch path {
      case "/state":
        body = bridgeState.stateJSON()
      case "/status":
        body = bridgeState.statusJSON()
      default:
        body = #"{"ok":false,"message":"not found"}"#
      }
      let statusLine = path == "/state" || path == "/status"
        ? "HTTP/1.1 200 OK\r\n"
        : "HTTP/1.1 404 Not Found\r\n"
      let headers =
        statusLine +
        "Content-Type: application/json\r\n" +
        "Access-Control-Allow-Origin: *\r\n" +
        "Access-Control-Allow-Methods: GET, OPTIONS\r\n" +
        "Access-Control-Allow-Headers: Content-Type\r\n" +
        "Cache-Control: no-store\r\n" +
        "Connection: close\r\n" +
        "Content-Length: \(body.utf8.count)\r\n\r\n"
      let response = method == "OPTIONS" ? statusLine + "Access-Control-Allow-Origin: *\r\nAccess-Control-Allow-Methods: GET, OPTIONS\r\nAccess-Control-Allow-Headers: Content-Type\r\nContent-Length: 0\r\n\r\n" : headers + body
      connection.send(content: response.data(using: .utf8), completion: .contentProcessed { _ in
        connection.cancel()
      })
    }
  }
}

guard let _ = HIDBridge() else {
  fputs("Failed to open IOHIDManager for 3Dconnexion devices.\n", stderr)
  exit(1)
}

guard let server = HTTPBridgeServer(port: bridgePort) else {
  fputs("Failed to start localhost HTTP bridge on port \(bridgePort).\n", stderr)
  exit(1)
}

server.start()
dispatchMain()

File: local-bridge/BoardCAD3DMouseBridge.c
#include <CoreFoundation/CoreFoundation.h>
#include <IOKit/hid/IOHIDManager.h>
#include <arpa/inet.h>
#include <math.h>
#include <netinet/in.h>
#include <pthread.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <unistd.h>

#define BRIDGE_PORT 8766
#define BRIDGE_VENDOR_ID 0x256F
#define BUTTON_CAPACITY 32

typedef struct {
  pthread_mutex_t mutex;
  double tx, ty, tz;
  double rx, ry, rz;
  bool buttons[BUTTON_CAPACITY];
  bool connected;
  char device_name[256];
  long long packet_count;
  double timestamp_ms;
} BridgeState;

static BridgeState g_state = {
  .mutex = PTHREAD_MUTEX_INITIALIZER,
  .tx = 0.0, .ty = 0.0, .tz = 0.0,
  .rx = 0.0, .ry = 0.0, .rz = 0.0,
  .connected = false,
  .device_name = "",
  .packet_count = 0,
  .timestamp_ms = 0.0
};

static double now_ms(void) {
  struct timeval tv;
  gettimeofday(&tv, NULL);
  return (double)tv.tv_sec * 1000.0 + (double)tv.tv_usec / 1000.0;
}

static void reset_axes_locked(BridgeState* state) {
  state->tx = 0.0;
  state->ty = 0.0;
  state->tz = 0.0;
  state->rx = 0.0;
  state->ry = 0.0;
  state->rz = 0.0;
  memset(state->buttons, 0, sizeof(state->buttons));
}

static void set_device_name_from_cfstring(char* dest, size_t dest_size, CFStringRef value) {
  if (!dest || dest_size == 0) return;
  dest[0] = '\0';
  if (!value) {
    snprintf(dest, dest_size, "%s", "3Dconnexion");
    return;
  }
  if (!CFStringGetCString(value, dest, (CFIndex)dest_size, kCFStringEncodingUTF8)) {
    snprintf(dest, dest_size, "%s", "3Dconnexion");
  }
}

static void bridge_set_device(IOHIDDeviceRef device, bool connected) {
  pthread_mutex_lock(&g_state.mutex);
  g_state.connected = connected;
  g_state.timestamp_ms = now_ms();
  if (device) {
    CFStringRef product = IOHIDDeviceGetProperty(device, CFSTR(kIOHIDProductKey));
    set_device_name_from_cfstring(g_state.device_name, sizeof(g_state.device_name), product);
  } else if (!connected) {
    snprintf(g_state.device_name, sizeof(g_state.device_name), "%s", "");
  }
  if (!connected) {
    reset_axes_locked(&g_state);
  }
  pthread_mutex_unlock(&g_state.mutex);
}

static void bridge_update_axis(const char* key, double value) {
  if (!key) return;
  if (value > 1.0) value = 1.0;
  if (value < -1.0) value = -1.0;
  pthread_mutex_lock(&g_state.mutex);
  if (strcmp(key, "tx") == 0) g_state.tx = value;
  else if (strcmp(key, "ty") == 0) g_state.ty = value;
  else if (strcmp(key, "tz") == 0) g_state.tz = value;
  else if (strcmp(key, "rx") == 0) g_state.rx = value;
  else if (strcmp(key, "ry") == 0) g_state.ry = value;
  else if (strcmp(key, "rz") == 0) g_state.rz = value;
  g_state.packet_count += 1;
  g_state.timestamp_ms = now_ms();
  pthread_mutex_unlock(&g_state.mutex);
}

static void bridge_update_button(int index, bool pressed) {
  if (index < 0 || index >= BUTTON_CAPACITY) return;
  pthread_mutex_lock(&g_state.mutex);
  g_state.buttons[index] = pressed;
  g_state.packet_count += 1;
  g_state.timestamp_ms = now_ms();
  pthread_mutex_unlock(&g_state.mutex);
}

static double normalize_axis(long value, long min, long max) {
  if (min < 0 && max > 0) {
    double denom = fabs((double)min) > fabs((double)max) ? fabs((double)min) : fabs((double)max);
    if (denom < 1e-9) return 0.0;
    return (double)value / denom;
  }
  double center = ((double)min + (double)max) * 0.5;
  double half = ((double)max - (double)min) * 0.5;
  if (fabs(half) < 1e-9) return 0.0;
  return ((double)value - center) / half;
}

static const char* axis_key(uint32_t usage_page, uint32_t usage) {
  if (usage_page != kHIDPage_GenericDesktop) return NULL;
  switch (usage) {
    case kHIDUsage_GD_X: return "tx";
    case kHIDUsage_GD_Y: return "ty";
    case kHIDUsage_GD_Z: return "tz";
    case kHIDUsage_GD_Rx: return "rx";
    case kHIDUsage_GD_Ry: return "ry";
    case kHIDUsage_GD_Rz: return "rz";
    default: return NULL;
  }
}

static void device_matching_callback(void* context, IOReturn result, void* sender, IOHIDDeviceRef device) {
  (void)context;
  (void)result;
  (void)sender;
  bridge_set_device(device, true);
}

static void device_removal_callback(void* context, IOReturn result, void* sender, IOHIDDeviceRef device) {
  (void)context;
  (void)result;
  (void)sender;
  (void)device;
  bridge_set_device(NULL, false);
}

static void input_value_callback(void* context, IOReturn result, void* sender, IOHIDValueRef value) {
  (void)context;
  (void)result;
  (void)sender;
  if (!value) return;
  IOHIDElementRef element = IOHIDValueGetElement(value);
  if (!element) return;
  uint32_t usage_page = IOHIDElementGetUsagePage(element);
  uint32_t usage = IOHIDElementGetUsage(element);
  if (usage_page == kHIDPage_Button) {
    bridge_update_button((int)usage, IOHIDValueGetIntegerValue(value) != 0);
    return;
  }
  const char* key = axis_key(usage_page, usage);
  if (!key) return;
  CFIndex raw = IOHIDValueGetIntegerValue(value);
  CFIndex min = IOHIDElementGetLogicalMin(element);
  CFIndex max = IOHIDElementGetLogicalMax(element);
  bridge_update_axis(key, normalize_axis((long)raw, (long)min, (long)max));
}

static void build_buttons_json(char* dest, size_t dest_size) {
  size_t offset = 0;
  int count = 0;
  offset += snprintf(dest + offset, dest_size > offset ? dest_size - offset : 0, "[");
  for (int i = 0; i < BUTTON_CAPACITY; i += 1) {
    if (!g_state.buttons[i]) continue;
    offset += snprintf(
      dest + offset,
      dest_size > offset ? dest_size - offset : 0,
      "%s%d",
      count > 0 ? "," : "",
      i
    );
    count += 1;
  }
  snprintf(dest + offset, dest_size > offset ? dest_size - offset : 0, "]");
}

static void build_state_json(char* dest, size_t dest_size) {
  char buttons_json[256];
  pthread_mutex_lock(&g_state.mutex);
  build_buttons_json(buttons_json, sizeof(buttons_json));
  snprintf(
    dest,
    dest_size,
    "{\"buttons\":%s,\"connected\":%s,\"deviceName\":\"%s\",\"packetCount\":%lld,\"rx\":%.8f,\"ry\":%.8f,\"rz\":%.8f,\"timestamp\":%.3f,\"tx\":%.8f,\"ty\":%.8f,\"tz\":%.8f}",
    buttons_json,
    g_state.connected ? "true" : "false",
    g_state.device_name,
    g_state.packet_count,
    g_state.rx, g_state.ry, g_state.rz,
    g_state.timestamp_ms,
    g_state.tx, g_state.ty, g_state.tz
  );
  pthread_mutex_unlock(&g_state.mutex);
}

static void build_status_json(char* dest, size_t dest_size) {
  pthread_mutex_lock(&g_state.mutex);
  snprintf(
    dest,
    dest_size,
    "{\"connected\":%s,\"deviceName\":\"%s\",\"ok\":true,\"packetCount\":%lld,\"port\":%d,\"timestamp\":%.3f}",
    g_state.connected ? "true" : "false",
    g_state.device_name,
    g_state.packet_count,
    BRIDGE_PORT,
    g_state.timestamp_ms
  );
  pthread_mutex_unlock(&g_state.mutex);
}

static void send_http_response(int client_fd, const char* status, const char* body) {
  char header[1024];
  size_t body_len = body ? strlen(body) : 0;
  int header_len = snprintf(
    header,
    sizeof(header),
    "%s\r\n"
    "Content-Type: application/json\r\n"
    "Access-Control-Allow-Origin: *\r\n"
    "Access-Control-Allow-Methods: GET, OPTIONS\r\n"
    "Access-Control-Allow-Headers: Content-Type\r\n"
    "Cache-Control: no-store\r\n"
    "Connection: close\r\n"
    "Content-Length: %zu\r\n\r\n",
    status,
    body_len
  );
  if (header_len > 0) {
    send(client_fd, header, (size_t)header_len, 0);
  }
  if (body && body_len > 0) {
    send(client_fd, body, body_len, 0);
  }
}

static void* http_server_thread(void* arg) {
  (void)arg;
  int server_fd = socket(AF_INET, SOCK_STREAM, 0);
  if (server_fd < 0) {
    perror("socket");
    return NULL;
  }
  int yes = 1;
  setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &yes, sizeof(yes));

  struct sockaddr_in addr;
  memset(&addr, 0, sizeof(addr));
  addr.sin_family = AF_INET;
  addr.sin_port = htons(BRIDGE_PORT);
  addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);

  if (bind(server_fd, (struct sockaddr*)&addr, sizeof(addr)) != 0) {
    perror("bind");
    close(server_fd);
    return NULL;
  }
  if (listen(server_fd, 16) != 0) {
    perror("listen");
    close(server_fd);
    return NULL;
  }

  printf("BoardCAD 3D mouse bridge listening on http://127.0.0.1:%d\n", BRIDGE_PORT);
  fflush(stdout);

  while (1) {
    int client_fd = accept(server_fd, NULL, NULL);
    if (client_fd < 0) continue;

    char request[4096];
    ssize_t received = recv(client_fd, request, sizeof(request) - 1, 0);
    if (received <= 0) {
      close(client_fd);
      continue;
    }
    request[received] = '\0';

    char method[16] = {0};
    char path[256] = {0};
    sscanf(request, "%15s %255s", method, path);

    if (strcmp(method, "OPTIONS") == 0) {
      send_http_response(client_fd, "HTTP/1.1 200 OK", "");
      close(client_fd);
      continue;
    }

    char body[1024];
    if (strcmp(path, "/state") == 0) {
      build_state_json(body, sizeof(body));
      send_http_response(client_fd, "HTTP/1.1 200 OK", body);
    } else if (strcmp(path, "/status") == 0) {
      build_status_json(body, sizeof(body));
      send_http_response(client_fd, "HTTP/1.1 200 OK", body);
    } else {
      snprintf(body, sizeof(body), "%s", "{\"ok\":false,\"message\":\"not found\"}");
      send_http_response(client_fd, "HTTP/1.1 404 Not Found", body);
    }
    close(client_fd);
  }

  close(server_fd);
  return NULL;
}

int main(void) {
  g_state.timestamp_ms = now_ms();

  IOHIDManagerRef manager = IOHIDManagerCreate(kCFAllocatorDefault, kIOHIDOptionsTypeNone);
  if (!manager) {
    fprintf(stderr, "Failed to create IOHIDManager.\n");
    return 1;
  }

  int vendor_id = BRIDGE_VENDOR_ID;
  CFNumberRef vendor_number = CFNumberCreate(kCFAllocatorDefault, kCFNumberIntType, &vendor_id);
  const void* match_keys[] = { CFSTR(kIOHIDVendorIDKey) };
  const void* match_values[] = { vendor_number };
  CFDictionaryRef match_dict = CFDictionaryCreate(
    kCFAllocatorDefault,
    match_keys,
    match_values,
    1,
    &kCFTypeDictionaryKeyCallBacks,
    &kCFTypeDictionaryValueCallBacks
  );
  IOHIDManagerSetDeviceMatching(manager, match_dict);
  IOHIDManagerRegisterDeviceMatchingCallback(manager, device_matching_callback, NULL);
  IOHIDManagerRegisterDeviceRemovalCallback(manager, device_removal_callback, NULL);
  IOHIDManagerRegisterInputValueCallback(manager, input_value_callback, NULL);
  IOHIDManagerScheduleWithRunLoop(manager, CFRunLoopGetMain(), kCFRunLoopDefaultMode);

  IOReturn result = IOHIDManagerOpen(manager, kIOHIDOptionsTypeNone);
  if (result != kIOReturnSuccess) {
    fprintf(stderr, "Failed to open IOHIDManager for 3Dconnexion devices.\n");
    if (match_dict) CFRelease(match_dict);
    if (vendor_number) CFRelease(vendor_number);
    CFRelease(manager);
    return 1;
  }

  pthread_t thread;
  if (pthread_create(&thread, NULL, http_server_thread, NULL) != 0) {
    fprintf(stderr, "Failed to start localhost HTTP bridge thread.\n");
    IOHIDManagerClose(manager, kIOHIDOptionsTypeNone);
    if (match_dict) CFRelease(match_dict);
    if (vendor_number) CFRelease(vendor_number);
    CFRelease(manager);
    return 1;
  }
  pthread_detach(thread);

  if (match_dict) CFRelease(match_dict);
  if (vendor_number) CFRelease(vendor_number);
  CFRunLoopRun();

  IOHIDManagerClose(manager, kIOHIDOptionsTypeNone);
  CFRelease(manager);
  return 0;
}
